jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/BuildImageStep.java

```diff
@@ -99,36 +99,43 @@
       Image<Layer> baseImage = NonBlockingSteps.get(pullBaseImageStep).getBaseImage();
       ContainerConfiguration containerConfiguration =
           buildConfiguration.getContainerConfiguration();
-      String buildTime =
-          containerConfiguration == null
-              ? Instant.EPOCH.toString()
-              : containerConfiguration.getCreationTime().toString();
 
-      // Base image layers/configuration passthrough
+      // Base image layers
       List<PullAndCacheBaseImageLayerStep> baseImageLayers =
           NonBlockingSteps.get(pullAndCacheBaseImageLayersStep);
       for (PullAndCacheBaseImageLayerStep pullAndCacheBaseImageLayerStep : baseImageLayers) {
         imageBuilder.addLayer(NonBlockingSteps.get(pullAndCacheBaseImageLayerStep));
       }
+
+      // Passthrough config and count non-empty history entries
+      int nonEmptyLayerCount = 0;
       for (HistoryObjectTemplate historyObject : baseImage.getHistory()) {
         imageBuilder.addHistory(historyObject);
+        if (!historyObject.isEmptyLayer()) {
+          nonEmptyLayerCount++;
+        }
       }
       imageBuilder.addEnvironment(baseImage.getEnvironment());
       imageBuilder.addLabels(baseImage.getLabels());
 
-      // Add history elements if base layers/history sizes don't match
-      if (baseImage.getHistory().size() < baseImageLayers.size()) {
-        int sizeDifference = baseImageLayers.size() - baseImage.getHistory().size();
-        for (int count = 0; count < sizeDifference; count++) {
-          imageBuilder.addHistory(new HistoryObjectTemplate(buildTime, "Jib", "jib"));
-        }
+      // Add history elements for non-empty layers that don't have one yet
+      int sizeDifference =
+          baseImageLayers.size() + buildAndCacheApplicationLayerSteps.size() - nonEmptyLayerCount;
+      for (int count = 0; count < sizeDifference; count++) {
+        imageBuilder.addHistory(
+            new HistoryObjectTemplate(
```

Comment:
Oh right, I just realized that this was passing a JSON template object into the `Image` builder. `Image` is intended to be a canonicalized Java object model of an image independent of the format (hence why we have `ImageToJsonTranslator`/`JsonToImageTranslator`). Making this distinction would allow for this image-building code not have to touch JSON templates at all.

Reply:
Is there a way to address this without making two nearly identical classes?

Reply:
Hmm actually I guess they might not be that identical.